### PR TITLE
Option IO throttling of the merger.

### PR DESCRIPTION
**Pull request by fulmicoton**

### CODE REVIEW

File: quickwit/quickwit-indexing/src/actors/merge_executor.rs [hunk 1]
```diff
@@ -233,6 +234,7 @@ async fn merge_split_directories(
         Box::new(MmapDirectory::open(output_path)?),
         ctx.progress().clone(),
         ctx.kill_switch().clone(),
+        Some(Byte::from_bytes(20_000_000)),
```
Comment (fulmicoton) [inline quickwit/quickwit-indexing/src/actors/merge_executor.rs:237]: need to fix

File: quickwit/quickwit-indexing/src/throttle.rs [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
+use std::sync::{Arc, Weak};
+use std::time::{Duration, Instant};
+
+use quickwit_common::metrics::IntCounter;
+pub struct Throttle {
```
Comment (guilload) [inline quickwit/quickwit-indexing/src/throttle.rs:25]: We might not need a custom implementation: the `governor` rate limiter wraps an `AtomicU64`, so cloneable with an `Arc` and has sync (`check*`) and async `until*` APIs.

Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/throttle.rs:25]: One thing I am scared with governor is that 
a) the docs only talks about requests / operations etc. and does not talk about using it to throttle a throughput
b) the logic that throttles is a loop that have tasks wait for the capacity to be available. I am not fond of that:
I don't know what limits the number of times a thread will have to wake up, I am not sure if there can be some wait of time here.
c) I don't understand how it works so I am not able to reason with it:
For instance, if I have two writers, sharing a rate limiter, and being throttled. One is writing 1kB at a time, when the other is doing it 10kb at a time, what throughput do they achieve?

Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/throttle.rs:25]: created an issue on github: https://github.com/antifuchs/governor/issues/154

Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/throttle.rs:25]: created an issue on github: https://github.com/antifuchs/governor/issues/154

Comment (fulmicoton) [reply quickwit/quickwit-indexing/src/throttle.rs:25]: https://github.com/tikv/async-speed-limit seems to do exactly what we need and more (can wrap async read/write directly).

**Comment by fulmicoton** [comment]
Switching to draft, will replace with async-speed-limit

**Comment by fulmicoton** [comment]
Closed for #2205

**Pull request closed without merging**